rayon-core/src/join/mod.rs

```diff
@@ -43,18 +45,35 @@ pub fn join<A, B, RA, RB>(oper_a: A, oper_b: B) -> (RA, RB)
           RA: Send,
           RB: Send
 {
-    registry::in_worker(|worker_thread| unsafe {
+    join_context(|_| oper_a(), |_| oper_b())
+}
+
+/// The `join_context` function is identical to `join`, except the closures
+/// have a parameter that provides context for the way the closure has been
+/// called, especially indicating whether they're executing on a different
+/// thread than where `join_context` was called.  This will occur if the second
+/// job is stolen by a different thread, or if `join_context` was called from
+/// outside the thread pool to begin with.
+pub fn join_context<A, B, RA, RB>(oper_a: A, oper_b: B) -> (RA, RB)
+    where A: FnOnce(FnContext) -> RA + Send,
+          B: FnOnce(FnContext) -> RB + Send,
```

Comment:
Hmm I think I didn't notice that it supports `Default`. Maybe it's overkill to worry about this stuff anyway.